migrations/account_type/migration.go

```diff
@@ -0,0 +1,237 @@
+	}
+}
+
+func (m *AccountTypeMigration) Migrate(
+	addressIterator migrations.AddressIterator,
+	reporter migrations.Reporter,
+) {
+	for {
+		address := addressIterator.NextAddress()
+		if address == common.ZeroAddress {
+			break
+		}
+
+		m.migrateTypeValuesInAccount(
+			address,
+			reporter,
+		)
+	}
+
+	err := m.storage.Commit(m.interpreter, false)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// migrateTypeValuesInAccount migrates `AuthAccount` and `PublicAccount` types in a given account
+// to the account reference type (&Account).
+func (m *AccountTypeMigration) migrateTypeValuesInAccount(
+	address common.Address,
+	reporter migrations.Reporter,
+) {
+
+	accountStorage := migrations.NewAccountStorage(m.storage, address)
+
+	accountStorage.ForEachValue(
+		m.interpreter,
+		common.AllPathDomains,
+		m.migrateValue,
+		reporter,
+	)
+}
+
+func (m *AccountTypeMigration) migrateValue(value interpreter.Value) (newValue interpreter.Value, updatedInPlace bool) {
+	return migrations.MigrateNestedValue(m.interpreter, value, m.migrateTypeValue)
+}
+
+func (m *AccountTypeMigration) migrateTypeValue(value interpreter.Value) (newValue interpreter.Value, updatedInPlace bool) {
+	if typeValue, ok := value.(interpreter.TypeValue); ok {
+		convertedType := m.maybeConvertAccountType(typeValue.Type)
+		if convertedType == nil {
+			return
+		}
+
+		return interpreter.NewTypeValue(nil, convertedType), true
+	}
+
+	return nil, false
+}
+
+func (m *AccountTypeMigration) maybeConvertAccountType(staticType interpreter.StaticType) interpreter.StaticType {
+	switch staticType := staticType.(type) {
+	case *interpreter.ConstantSizedStaticType:
+		convertedType := m.maybeConvertAccountType(staticType.Type)
+		if convertedType != nil {
+			return interpreter.NewConstantSizedStaticType(nil, convertedType, staticType.Size)
+		}
+
+	case *interpreter.VariableSizedStaticType:
+		convertedType := m.maybeConvertAccountType(staticType.Type)
+		if convertedType != nil {
+			return interpreter.NewVariableSizedStaticType(nil, convertedType)
+		}
+
+	case *interpreter.DictionaryStaticType:
+		convertedKeyType := m.maybeConvertAccountType(staticType.KeyType)
+		convertedValueType := m.maybeConvertAccountType(staticType.ValueType)
+		if convertedKeyType != nil && convertedValueType != nil {
+			return interpreter.NewDictionaryStaticType(nil, convertedKeyType, convertedValueType)
+		}
+		if convertedKeyType != nil {
+			return interpreter.NewDictionaryStaticType(nil, convertedKeyType, staticType.ValueType)
+		}
+		if convertedValueType != nil {
+			return interpreter.NewDictionaryStaticType(nil, staticType.KeyType, convertedValueType)
+		}
+
+	case *interpreter.CapabilityStaticType:
+		convertedBorrowType := m.maybeConvertAccountType(staticType.BorrowType)
+		if convertedBorrowType != nil {
+			return interpreter.NewCapabilityStaticType(nil, convertedBorrowType)
+		}
+
+	case *interpreter.IntersectionStaticType:
+		// Nothing to do. Inner types can only be interfaces.
```

Comment:
Maybe still migrate `LegacyType`, the previous restricted type

Reply:
I'll add this in a follow-up PR